tests/subsys/logging/log_switch_format/src/main.c

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/zephyr.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/printk.h>
+#include <zephyr/logging/log_ctrl.h>
+#include <zephyr/logging/log_output.h>
+#include <zephyr/logging/log_output_dict.h>
+#include <ztest.h>
+
+#define LOG_MODULE_NAME log_switch_format
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL_ERR);
+
+extern size_t log_format_table_size(void);
+
+void log_msgs(void)
+{
+	/* standard print */
+	LOG_ERR("Error message example.");
+
+	/* raw string */
+	printk("hello sys-t on board %s\n", CONFIG_BOARD);
+
+#if CONFIG_LOG_MODE_DEFERRED
+	/*
+	 * When deferred logging is enabled, the work is being performed by
+	 * another thread. This k_sleep() gives that thread time to process
+	 * those messages.
+	 */
+
+	k_sleep(K_TICKS(10));
+#endif
+}
+
+void test_log_switch_format_success_case(void)
+{
+	const struct log_backend *backend;
+	uint32_t log_type;
+
+	log_msgs();
+
+	log_type = LOG_OUTPUT_TEXT;
+	backend = log_format_set_all_active_backends(log_type);
+
+	zassert_is_null(backend, "Unexpected failure in switching log format");
+
+	log_msgs();
+
+	log_type = LOG_OUTPUT_SYST;
+	backend = log_format_set_all_active_backends(log_type);
+
+	zassert_is_null(backend, "Unexpected failure in switching log format");
+
+	log_msgs();
+
+	log_type = LOG_OUTPUT_TEXT;
+	backend = log_format_set_all_active_backends(log_type);
+
+	zassert_is_null(backend, "Unexpected failure in switching log format");
+
+	/* raw string */
+	printk("SYST Sample Execution Completed\n");
+
+}
+
+void test_log_switch_format_set(void)
+{
+	const char *backend_name;
+	const struct log_backend *backend;
+	int ret;
+	uint32_t log_type;
+
+	log_type = LOG_OUTPUT_TEXT;
+	backend_name = "not_exists";
+
+	backend = log_backend_get_by_name(backend_name);
+	zassert_is_null(backend, "Backend unexpectedly found");
+
+	ret = log_backend_format_set(backend, log_type);
+
+	zassert_equal(ret, -EINVAL, "Expected -EINVAL, Got %d\n", ret);
+
+	backend_name = CONFIG_LOG_BACKEND_DEFAULT;
+	backend = log_backend_get_by_name(backend_name);
+
+	zassert_not_null(backend, "Backend not found");
+
+	log_type = log_format_table_size() + 1;
+	ret = log_backend_format_set(backend, log_type);
+
+	zassert_equal(ret, -EINVAL, "Log type not supported, Invalid value returned");
+
+}
+
+void test_log_switch_format_set_all_active_backends(void)
+{
+	size_t log_type = log_format_table_size() + 1;
+	const struct log_backend *backend;
+
+	backend = log_format_set_all_active_backends(log_type);
+
+	zassert_not_null(backend, "Not all active backends have switched logging formats");
```

Comment:
I guess the message here should be different, no? What is unexpected here is that any backend accept this change, right? I wonder if there's a way to ensure that the very first backend failed...

Reply:
The unexpected behavior is when from all the active backends, any specific backend does not change the logging format. If all the backends successfully change the log format dynamically then "backend" would contain NULL , otherwise it will be a pointer to the very first backend that failed.

Reply:
But isn't in this case setting an invalid backend? That's my impression from line 101. Also, you are checking that this change fails, as is `zassert_non_null`.
So, at least, the message doesn't make much sense. But I still think that if there was a way to ensure that all backends fail it would be better, no? We don't want any backend quietly accepting an invalid log format...

Reply:
The new message is much better =D

About the feature, maybe I'm misunderstanding it. What I'm afraid of is the following scenario:
 - Let `A`, `B`, `C` and `D` be backends;
 - Let `α`, `β`, `γ` be log formats;
 Now, if we try to set log format to `δ`, this function will return one of [`A`, `B`, `C`, `D`], meaning that the change failed. Assuming the backends are checked in that order, it will return `A`. Suppose backend `A` is buggy and accepts the invalid log format, the function will return `B`. If someone only enables backend `A`, buggy behaviour will happen, but we would never caught it.
Is a scenario like this possible? Maybe not, and I'm talking gibberish. But if it is, we may want to test for it. Or may not, if it's not worth the effort.

Reply:
@edersondisouza  Okay. I am kind of getting what your point is. So the backend can be buggy in number of ways. But when we say backend is buggy in context with switching of logging formats at runtime, it will not matter because setting/switching of logging formats functionality has been separated from the backends functionality itself. To swtich the logging formats the required Kconfig options should also be set and the log format δ passed by the user should also exist as an entry in the function pointer table log_format_func_t format_table. Which means that a backend cannot accept any invalid log format. Hope this makes sense. Let me know.

Reply:
Oh I see. So, this type of error (invalid log format) is caught by the core logging (`include/zephyr/logging/log_backend.h`), so this test should be enough. Thanks for the explanation!